### PR TITLE
refactor: cm details/math ints => uints, documentation

### PR DESCRIPTION
**Pull request by dsshap**

### CODE REVIEW

**Comment by dsshap** [comment]
@hashnoteleo I believe you've been reviewing the branch for changes related to `ints => uints`. Making this official.
Also, i have been adding inline/signature documentation to the math.sol functions. I loosely incorporated comments from the notebooks, if you feel like there is areas that could use some additional comments please feel free to add.

@antoncoding the rest of the changes are for your review.

File: src/core/engines/cross-margin/CrossMarginLib.sol [hunk 1]
```diff
@@ -149,11 +149,10 @@ library CrossMarginLib {
         } else account.longs[index].amount = newLongAmount;
     }
 
-    ///@dev Settles the accounts short calls and puts, reserving collateral for ITM options
+    ///@dev Settles the accounts longs and shorts
     ///@param account CrossMarginAccount memory that will be updated in-place
```
Comment (antoncoding) [inline src/core/engines/cross-margin/CrossMarginLib.sol:153]: this comment seems wrong: we're updating account storage directly

File: src/core/engines/cross-margin/CrossMarginLib.sol [hunk 2]
```diff
@@ -162,6 +161,9 @@ library CrossMarginLib {
         shortPayouts = _settleShorts(grappa, account);
     }
 
+    ///@dev Settles the accounts longs, adding collateral to balances
+    ///@param grappa interface to settle long options in a batch call
+    ///@param account CrossMarginAccount memory that will be updated in-place
```
Comment (antoncoding) [inline src/core/engines/cross-margin/CrossMarginLib.sol:166]: this comment seems wrong: we're updating account storage directly

File: src/core/engines/cross-margin/CrossMarginLib.sol [hunk 3]
```diff
@@ -199,6 +201,9 @@ library CrossMarginLib {
         }
     }
 
+    ///@dev Settles the accounts shorts, reserving collateral for ITM options
+    ///@param grappa interface to get short option payouts in a batch call
+    ///@param account CrossMarginAccount memory that will be updated in-place
```
Comment (antoncoding) [inline src/core/engines/cross-margin/CrossMarginLib.sol:206]: same

File: src/core/engines/cross-margin/CrossMarginMath.sol [hunk 4]
```diff
@@ -99,14 +108,15 @@ library CrossMarginMath {
 
     /**
      * @notice get minimum collateral
+     * @dev detail is composed of positions with the same underlying + strike + collateral + expiry
      * @param _detail margin details
-     * @return cashNeeded with {collateral asset's} decimals
+     * @return numeraireNeeded with {collateral asset's} decimals
      * @return underlyingNeeded with {underlying asset's} decimals
      */
```
Comment (antoncoding) [inline src/core/engines/cross-margin/CrossMarginMath.sol:115]: what is `numeraireNeeded` vs `underlyingNeeded`? 
It's unclear which asset they are denominated in.
If collateral is underlying, does it return the same value?

Comment (antoncoding) [reply src/core/engines/cross-margin/CrossMarginMath.sol:115]: Seems like `numeraireNeeded` should be denominated in USDC?

File: src/core/engines/cross-margin/CrossMarginMath.sol [hunk 5]
```diff
@@ -173,31 +182,50 @@ library CrossMarginMath {
...
+     * @notice get numeraire and underlying needed to fully collateralize
+     * @dev calculates left side and right side of the payout profile
+     * @param putStrikes array of put option strikes
+     * @param putWeights amount of options at a given strike
+     * @param hasCalls has call options
+     * @param pois are the strikes we are evaluating
+     * @param payouts are the payouts at a given strike
+     * @return numeraireNeeded with {collateral asset's} decimals
+     * @return underlyingNeeded with {underlying asset's} decimals
+     */
```
Comment (antoncoding) [inline src/core/engines/cross-margin/CrossMarginMath.sol:195]: same comment here: Not sure what would happen if collateral asset == underlying asset

**Comment by antoncoding** [comment]
Overall, I think the structure is assuming that the collateral is either strike asset or underlying
It's a bit confusing that the term used is `numeraire` and `underlying`, and the term `numeraire` is assumed to be collateral all most of the time (but instead it should be representing the strike asset)

**Comment by dsshap** [comment]
@antoncoding where do you see the structure assuming collateral is either strike or underlying?

File: src/core/engines/cross-margin/CrossMarginLib.sol (hunk 1 above)
Comment (dsshap) [reply src/core/engines/cross-margin/CrossMarginLib.sol:153]: Yep - thats correct. This was copied over from FullMarginLib. It'll be worth opening a dedicated PR for it.

File: src/core/engines/cross-margin/CrossMarginMath.sol (hunk 4 above)
Comment (dsshap) [reply src/core/engines/cross-margin/CrossMarginMath.sol]: Numeraire is in fact the better term here. 

Grappa is overloading the term `strike` to also describe the asset. 

In ETH / USDC, underlying is ETH and numeraire is USDC.

> If collateral is underlying, does it return the same value?
It certainly would not return the same value. Calls are collateralized in underlying, puts are collateralized in numeraire. In CrossMarginDetail we are collecting all the puts and calls that have the same underlying, numeraire (strike) asset, and expiry. Based on this combination we can cross margin the options.

If there are no put options, then the whole thing is assumed to be collateralized in underlying. If there is numeraire required to be posted (example call spread) then that amount is converted into equivalent underlying.

Comment (dsshap) [reply src/core/engines/cross-margin/CrossMarginMath.sol]: I did, in the last commit, update the CrossMarginDetail struct to reflect what it actually should have been this whole time.

File: src/core/engines/cross-margin/CrossMarginMath.sol (hunk 5 above)
Comment (dsshap) [reply src/core/engines/cross-margin/CrossMarginMath.sol]: see above

**Comment by dsshap** [comment]
@antoncoding any thoughts on https://github.com/antoncoding/grappa/pull/160#issuecomment-1324156227?

**Comment by antoncoding** [comment]
> @antoncoding any thoughts on [#160 (comment)](https://github.com/antoncoding/grappa/pull/160#issuecomment-1324156227)?

I stood by my old statement to avoid that duplicated check. I think gas a side, a clean implementation should not have duplicate checks.

**Comment by hashnoteleo** [comment]
Thought a little more about this. Is there no future path that may enter
the code and hit one check but not the other? It looks like not in the
current use case, but is there a case one may want to build something that
calls _processDetails without _verifyInputs?

Cheers,
L

File: src/core/engines/cross-margin/CrossMarginMath.sol [hunk 6]
```diff
@@ -211,13 +239,23 @@ library CrossMarginMath {
...
 
+    /**
+     * @notice generating points of interest (strikes)
+     * @dev adding a point left of left most strike (puts) and a right point of right most strike (call)
+     * @param strikes array of shorts and longs
+     * @param numOfPuts number of puts
+     * @return pois array of point-of-interests (aka strikes)
+     */
     function _createPois(uint256[] memory strikes, uint256 numOfPuts) internal pure returns (uint256[] memory pois) {
         uint256 epsilon = strikes.min() / 10;
```
Comment (antoncoding) [inline src/core/engines/cross-margin/CrossMarginMath.sol:260]: you should cache `strikes.min()` here, so you don't have to re-calculate it later.
Even better if you can use 1 loop to find min and max

**Review by antoncoding — APPROVED** [review]
Comments looking good!

File: src/core/engines/cross-margin/CrossMarginMath.sol [hunk 7]
```diff
@@ -137,14 +150,10 @@ library CrossMarginMath {
                 _detail.putStrikes.length > 0
```
[the remark this answers is not in the document]
Comment (dsshap) [reply src/core/engines/cross-margin/CrossMarginMath.sol:150]: Yep, good catch

File: src/core/engines/cross-margin/CrossMarginMath.sol [hunk 8]
```diff
@@ -173,31 +182,50 @@ library CrossMarginMath {
...
+        uint256[] memory putStrikes,
+        int256[] memory putWeights,
+        bool hasCalls,
         uint256[] memory pois,
         int256[] memory payouts
-    ) internal pure returns (int256 cashNeeded, int256 underlyingNeeded) {
-        bool hasCalls = _detail.callStrikes.length > 0;
-        bool hasPuts = _detail.putStrikes.length > 0;
+    ) internal pure returns (uint256 numeraireNeeded, uint256 underlyingNeeded) {
+        bool hasPuts = putStrikes.length > 0;
```
[the remark this answers is not in the document]
Comment (dsshap) [reply src/core/engines/cross-margin/CrossMarginMath.sol:203]: I dont want to pass the array of calls strikes just for symmetry at the cost increased gas. Also, no downstream functions need those strikes extracted.

**Pull request merged**